### PR TITLE
[TLS 1.3] Support signature_algorithm_cert Extension

### PR DESCRIPTION
**Pull request by reneme**
## Pull Request Dependencies

* ~~#3053~~
* ~~#3081~~
* ~~#3186~~

## Description

This introduces support for [the `signature_algorithm_cert` extension](https://www.rfc-editor.org/rfc/rfc8446.html#section-4.2.3) introduced with TLS 1.3. With that, one can create a finer-grained signature algorithm policy; independently specifying acceptable algorithms for signatures in X.509 certificates and in the TLS protocol itself (i.e. in the `CertificateVerify` message).

To communicate the peer's restrictions to the application, `CredentialsManager::find_cert_chain()` (and friends) will now additionally get a vector of `AlgorithmIdentifier`s specifying which algorithms are acceptable.

Note that also the TLS 1.2 implementation can process this new extension (as suggested by RFC 8446).

## New policy: `tls_acceptable_certificate_signature_schemes()`

By default, this returns `std::nullopt` meaning that the restrictions from the existing `::tls_acceptable_signature_schemes()` should also apply for signatures in certificates. If anything other than `std::nullopt` is returned, clients will add an explicit `signature_algorithm_cert` extension to the Client Hello; servers likewise in the `CertificateRequest` message.

### CODE REVIEW

File: src/lib/tls/tls_signature_scheme.cpp
Comment (reneme) [inline src/lib/tls/tls_signature_scheme.cpp:328]: This is somewhat a duplicate of the implementations of `EMSA::config_for_x509()` scattered around the different EMSAs. I'm not quite happy with that, to be honest. Simply instantiating the respective EMSA and then actually using `::config_for_x509()` doesn't work out-of-the-box, as it would currently require to pass a respective private key. I think it would be helpful to abstract both use cases and centralize this somehow.

Anyway: As is, this is fairly inefficient, as the `AlgorithmIdentifier`s are re-generated for each invocation. It would be an interesting exercise to `constexpr` the heck out of this. 🤡

I'm open for better ideas. :)

**Comment by reneme** [comment]
Whoops, that almost broke the code examples. @lieser thanks for the CI job checking their successful compilation. 😅

File: src/lib/tls/tls_signature_scheme.cpp
Comment (randombit) [reply src/lib/tls/tls_signature_scheme.cpp:328]: Yeah pretty ugly. Also we recompute the PSS params each time when in practice it will always be SHA-{256,512} :sob: 

Looking at `config_for_x509` ... why the heck does this interface take a `Private_Key`? It could easily get away with a `Public_Key`? (Not that this change would address the issue here, but still kind of odd).

Anyway this falls in the category of definitely not great and pointing to a place where some refactoring time is needed but doesn't IMO need to block the PR merging.

File: src/lib/tls/tls_signature_scheme.cpp
Comment (randombit) [inline src/lib/tls/tls_signature_scheme.cpp:324]: Hold on a minute - do we still support DSA ciphersuites?

It appears we do?? Or at least there is something left over after #2505 which I had though removed DSA entirely...

A 3.0 release seems like a very opportune time to drop all support for DSA from TLS, if we haven't already ...

Comment (randombit) [reply src/lib/tls/tls_signature_scheme.cpp:324]: OK it appears the only real usage is in the `Signature_Scheme::Code` - will open a PR

Comment (randombit) [reply src/lib/tls/tls_signature_scheme.cpp]: Removed in #3143

**Review by randombit — APPROVED** [review]
Needs a rebase in order to (hopefully) make use of #3142, but otherwise lgtm.

File: src/lib/tls/tls_signature_scheme.cpp
Comment (reneme) [inline src/lib/tls/tls_signature_scheme.cpp:254]: This is now using `EMSA::config_for_x509()` instead of [copy-pasting most of its scattered implementations](https://github.com/randombit/botan/pull/3125#discussion_r1049941524).

File: src/lib/tls/tls_signature_scheme.cpp
Comment (reneme) [reply src/lib/tls/tls_signature_scheme.cpp]: After merging #3142, this now became: https://github.com/randombit/botan/pull/3125#discussion_r1070925009
Thanks for the help in adapting `config_for_x509()` to make it suitable for this use case!

File: src/lib/tls/tls_signature_scheme.cpp
Comment (reneme) [reply src/lib/tls/tls_signature_scheme.cpp:254]: The `config_for_x509()` implementations check that the passed hash function name is equal to the hash function used in the EMSA. This is done by string comparison (e.g. from emsa1.cpp):

```c++
if(cert_hash_name != m_hash->name())
   throw Invalid_Argument(...);
```

The library can deal with different string-identifiers for hash functions (e.g "SHA-1", "SHA1" and "SHA-160"). But `m_hash->name()` usually returns a hard-coded identifier. I tried dealing with that using the "SHA-1" to "SHA-160" mapping above. However, this PR's macOS builds fail, because the software-implementation (sha1.cpp) returns "SHA-160" when invoking `Hash_Function::name()` and the CommonCrypto provider (`CommonCrypto_HashFunction`) that might or might not be enabled on macOS returns "SHA-1". 😨

In my opinion, these string comparisons are a generic footgun that will cause recurring issues. 😞 Especially when they force us to scatter such mappings around the code base. Lets try to centralize the string-identifier mappings in the algorithm implementations? @randombit

This could be done by requiring a static method `::implements(std::string_view algo_name)` in each concrete implementation. Now, that implementations "know" all their string-identifiers, a `virtual bool Hash_Function::is(std::string_view algo_name)` could allow convenient polymorphic checks as in `config_for_x509()`. Even the `Hash_Function::create()` factory could benefit from `::implements()` and stop duplicating the string-identifiers. See [this godbolt for a sketch](https://godbolt.org/z/MGxEfjj3M).

One might even consider to auto-generate the `Hash_Function::create()` code based on self-registering sub-classes. But that might take it too far.

Obviously, getting rid of the ambiguous string-identifiers would be even better, but I doubt the massive public API change would be worth it.

Comment (randombit) [reply src/lib/tls/tls_signature_scheme.cpp:254]: > However, this PR's macOS builds fail, because the software-implementation (sha1.cpp) returns "SHA-160" when invoking Hash_Function::name() and the CommonCrypto provider (CommonCrypto_HashFunction) that might or might not be enabled on macOS returns "SHA-1"

Can you take a look at #3184?

> In my opinion, these string comparisons are a generic footgun that will cause recurring issues. disappointed Especially when they force us to scatter such mappings around the code base. Lets try to centralize the string-identifier mappings in the algorithm implementations?

They are. It was one of the big design mistakes.

> One might even consider to auto-generate the Hash_Function::create() code based on self-registering sub-classes.

This used to be how it worked, actually. But that approach ran into weird problems with static libraries, and ended up having to be removed (in https://github.com/randombit/botan/pull/279)

> This could be done by requiring a static method ::implements(std::string_view algo_name) in each concrete implementation. Now, that implementations "know" all their string-identifiers, a virtual bool Hash_Function::is(std::string_view algo_name) could allow convenient polymorphic checks as in config_for_x509().

This might work but I wonder how this handles parameterizations such as output length or personalization flags. Quick example

```
auto hash = HashFunction::create("BLAKE2b(384)");
assert!(hash->is("BLAKE2b")); // true?
assert!(hash->is("BLAKE2b(384)")); // true?
```

As far as I can recall the main problematic case for this is SHA-1, largely because we use the idiosyncratic/non-standard "SHA-160" name. Maybe the real fix here is to just remove that entirely as a known name, and relnote it? It's not even a big compatibility issue since "SHA-1" always worked.

> Obviously, getting rid of the ambiguous string-identifiers would be even better, but I doubt the massive public API change would be worth it.

We could consider having an enum approach and then implementing the string method on top of that, but I agree that we cannot really remove the string-based interface at this point since it's very widely baked into how the CLI, the tests, the FFI layer, etc all work.

Comment (reneme) [reply src/lib/tls/tls_signature_scheme.cpp:254]: > This might work but I wonder how this handles parameterizations such as output length or personalization flags.

Fair point. A static `BLAKE2b::implements("BLAKE2b(384)")` should return `true`. But a specific instance of it probably shouldn't.

> Maybe the real fix here is to just remove that entirely as a known name, and relnote it?

I tend to agree. Given that the string-identifiers are such a footgun its probably a good idea to make them as dumb as possible. As in: Don't allow for ambiguities during construction.

> We could consider having an enum approach [...]

Though, I don't see how parameterization would work well with enums either. `Hash_Types::BLAKE2b_384`? That feels rigid to me.

Would it make sense to use `class AlgorithmIdentifier` for those use cases -- as its name suggests? De facto this is already implemented in the specific algorithm implementations. As a library user I always found it inconvenient/confusing to construct `AlgorithmIdentifier` out of thin air, but that might be fixable by some user-friendly API.

**Comment by reneme** [comment]
Rebased and cleaned up after #3186 was merged. This should be ready for prime time.

**Pull request merged**
